File: tools/webui/protocol.js
export const PROTOCOL_VERSION = 1;
export const MAGIC = [0xa5, 0x5a];
export const MAX_PAYLOAD_SIZE = 240;
export const FRAME_OVERHEAD = 11;
export const MAX_FRAME_SIZE = FRAME_OVERHEAD + MAX_PAYLOAD_SIZE;

export const CommandType = Object.freeze({
  Ping: 0x01,
  GetInfo: 0x02,
  GetCaps: 0x03,
  KeyDown: 0x10,
  KeyUp: 0x11,
  KeyTap: 0x12,
  TypeAscii: 0x13,
  MouseMoveRel: 0x20,
  MouseButtonDown: 0x21,
  MouseButtonUp: 0x22,
  MouseClick: 0x23,
  MouseWheel: 0x24,
  WaitMs: 0x30,
  BatchBegin: 0x40,
  BatchEnd: 0x41,
  StopAll: 0x7f,
  Ack: 0x80,
  Nack: 0x81,
  Status: 0x82,
  Busy: 0x83,
});

export const CommandName = Object.freeze(
  Object.fromEntries(Object.entries(CommandType).map(([name, value]) => [value, name])),
);

export class DecodeError extends Error {
  constructor(code, message) {
    super(message);
    this.name = "DecodeError";
    this.code = code;
  }
}

export function encodeFrame(sequence, commandType, payload = new Uint8Array()) {
  const body = toBytes(payload);
  if (body.length > MAX_PAYLOAD_SIZE) {
    throw new RangeError(`payload is ${body.length} bytes, max is ${MAX_PAYLOAD_SIZE}`);
  }

  const frame = new Uint8Array(FRAME_OVERHEAD + body.length);
  frame[0] = MAGIC[0];
  frame[1] = MAGIC[1];
  frame[2] = PROTOCOL_VERSION;
  frame[3] = 0;
  writeU16(frame, 4, sequence);
  frame[6] = commandType;
  writeU16(frame, 7, body.length);
  frame.set(body, 9);

  const crc = crc16CcittFalse(frame.subarray(2, 9 + body.length));
  writeU16(frame, 9 + body.length, crc);
  return frame;
}

export function decodeFrame(input) {
  const frame = toBytes(input);
  if (frame.length < FRAME_OVERHEAD) {
    throw new DecodeError("too_short", "frame is too short");
  }
  if (frame[0] !== MAGIC[0] || frame[1] !== MAGIC[1]) {
    throw new DecodeError("bad_magic", "frame magic does not match");
  }

  const payloadLength = readU16(frame, 7);
  if (payloadLength > MAX_PAYLOAD_SIZE) {
    throw new DecodeError("payload_too_long", "payload is too long");
  }

  const expectedLength = FRAME_OVERHEAD + payloadLength;
  if (frame.length !== expectedLength) {
    throw new DecodeError("length_mismatch", "frame length does not match payload length");
  }

  const crcOffset = 9 + payloadLength;
  const expectedCrc = readU16(frame, crcOffset);
  const actualCrc = crc16CcittFalse(frame.subarray(2, crcOffset));
  if (expectedCrc !== actualCrc) {
    throw new DecodeError("bad_crc", "frame CRC check failed");
  }

  return {
    version: frame[2],
    flags: frame[3],
    sequence: readU16(frame, 4),
    commandType: frame[6],
    payload: frame.slice(9, crcOffset),
  };
}

export function extractFrames(buffer) {
  const frames = [];
  let offset = 0;

  while (buffer.length - offset >= 2) {
    if (buffer[offset] !== MAGIC[0] || buffer[offset + 1] !== MAGIC[1]) {
      offset += 1;
      continue;
    }

    if (buffer.length - offset < 9) {
      break;
    }

    const payloadLength = readU16(buffer, offset + 7);
    const frameLength = FRAME_OVERHEAD + payloadLength;
    if (frameLength > MAX_FRAME_SIZE) {
      offset += 2;
      continue;
    }
    if (buffer.length - offset < frameLength) {
      break;
    }

    frames.push(buffer.slice(offset, offset + frameLength));
    offset += frameLength;
  }

  return {
    frames,
    remaining: buffer.slice(offset),
  };
}

export function expectedResponseType(commandType) {
  if (commandType === CommandType.GetInfo || commandType === CommandType.GetCaps) {
    return CommandType.Status;
  }
  return CommandType.Ack;
}

export function crc16CcittFalse(data) {
  let crc = 0xffff;
  for (const byte of data) {
    crc ^= byte << 8;
    for (let i = 0; i < 8; i += 1) {
      if ((crc & 0x8000) !== 0) {
        crc = ((crc << 1) ^ 0x1021) & 0xffff;
      } else {
        crc = (crc << 1) & 0xffff;
      }
    }
  }
  return crc;
}

export function bytesToHex(bytes) {
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0").toUpperCase()).join(" ");
}

export function asciiPayload(text) {
  const out = new Uint8Array(text.length);
  for (let i = 0; i < text.length; i += 1) {
    const code = text.charCodeAt(i);
    if (code > 0x7f) {
      throw new Error("TYPE_ASCII only accepts ASCII text");
    }
    out[i] = code;
  }
  return out;
}

export function i16PairPayload(dx, dy) {
  const payload = new Uint8Array(4);
  writeI16(payload, 0, dx);
  writeI16(payload, 2, dy);
  return payload;
}

export function u32Payload(value) {
  const payload = new Uint8Array(4);
  payload[0] = (value >>> 24) & 0xff;
  payload[1] = (value >>> 16) & 0xff;
  payload[2] = (value >>> 8) & 0xff;
  payload[3] = value & 0xff;
  return payload;
}

export function keyPayload(combo) {
  return new Uint8Array([combo.modifier, combo.keycode]);
}

export function bytePayload(value) {
  return new Uint8Array([value & 0xff]);
}

function toBytes(input) {
  if (input instanceof Uint8Array) {
    return input;
  }
  if (input instanceof ArrayBuffer) {
    return new Uint8Array(input);
  }
  if (Array.isArray(input)) {
    return new Uint8Array(input);
  }
  throw new TypeError("expected bytes");
}

function readU16(bytes, offset) {
  return (bytes[offset] << 8) | bytes[offset + 1];
}

function writeU16(bytes, offset, value) {
  bytes[offset] = (value >>> 8) & 0xff;
  bytes[offset + 1] = value & 0xff;
}

function writeI16(bytes, offset, value) {
  const signed = value < 0 ? 0x10000 + value : value;
  writeU16(bytes, offset, signed & 0xffff);
}

File: sdk/python/rp2350_hid_bridge/protocol.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

MAGIC = b"\xA5\x5A"
PROTOCOL_VERSION = 1
MAX_PAYLOAD_SIZE = 240
FRAME_OVERHEAD = 11
MAX_FRAME_SIZE = FRAME_OVERHEAD + MAX_PAYLOAD_SIZE


class CommandType(IntEnum):
    PING = 0x01
    GET_INFO = 0x02
    GET_CAPS = 0x03
    KEY_DOWN = 0x10
    KEY_UP = 0x11
    KEY_TAP = 0x12
    TYPE_ASCII = 0x13
    MOUSE_MOVE_REL = 0x20
    MOUSE_BUTTON_DOWN = 0x21
    MOUSE_BUTTON_UP = 0x22
    MOUSE_CLICK = 0x23
    MOUSE_WHEEL = 0x24
    WAIT_MS = 0x30
    BATCH_BEGIN = 0x40
    BATCH_END = 0x41
    STOP_ALL = 0x7F
    ACK = 0x80
    NACK = 0x81
    STATUS = 0x82
    BUSY = 0x83


class DecodeError(ValueError):
    pass


@dataclass(frozen=True)
class Frame:
    version: int
    flags: int
    sequence: int
    command_type: CommandType | int
    payload: bytes


@dataclass(frozen=True)
class Response:
    command_type: CommandType | int
    payload: bytes
    sequence: int


def encode_frame(
    sequence: int,
    command_type: CommandType | int,
    payload: bytes = b"",
    version: int = PROTOCOL_VERSION,
) -> bytes:
    if len(payload) > MAX_PAYLOAD_SIZE:
        raise ValueError(f"payload is {len(payload)} bytes, max is {MAX_PAYLOAD_SIZE}")

    header = bytearray(FRAME_OVERHEAD + len(payload))
    header[0:2] = MAGIC
    header[2] = version & 0xFF
    header[3] = 0
    header[4:6] = int(sequence).to_bytes(2, "big")
    header[6] = int(command_type) & 0xFF
    header[7:9] = len(payload).to_bytes(2, "big")
    header[9 : 9 + len(payload)] = payload

    crc = crc16_ccitt_false(header[2 : 9 + len(payload)])
    header[9 + len(payload) : 11 + len(payload)] = crc.to_bytes(2, "big")
    return bytes(header)


def decode_frame(data: bytes | bytearray) -> Frame:
    frame = bytes(data)
    if len(frame) < FRAME_OVERHEAD:
        raise DecodeError("frame is too short")
    if frame[0:2] != MAGIC:
        raise DecodeError("bad magic")

    payload_len = int.from_bytes(frame[7:9], "big")
    if payload_len > MAX_PAYLOAD_SIZE:
        raise DecodeError("payload too long")

    expected_len = FRAME_OVERHEAD + payload_len
    if len(frame) != expected_len:
        raise DecodeError("length mismatch")

    crc_offset = 9 + payload_len
    expected_crc = int.from_bytes(frame[crc_offset : crc_offset + 2], "big")
    actual_crc = crc16_ccitt_false(frame[2:crc_offset])
    if expected_crc != actual_crc:
        raise DecodeError("bad crc")

    command_raw = frame[6]
    try:
        command_type: CommandType | int = CommandType(command_raw)
    except ValueError:
        command_type = command_raw

    return Frame(
        version=frame[2],
        flags=frame[3],
        sequence=int.from_bytes(frame[4:6], "big"),
        command_type=command_type,
        payload=frame[9:crc_offset],
    )


def expected_response_type(command_type: CommandType | int) -> CommandType:
    if command_type in (CommandType.GET_INFO, CommandType.GET_CAPS):
        return CommandType.STATUS
    return CommandType.ACK


def crc16_ccitt_false(data: bytes | bytearray) -> int:
    crc = 0xFFFF
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc


def ascii_payload(text: str) -> bytes:
    try:
        return text.encode("ascii")
    except UnicodeEncodeError as exc:
        raise ValueError("TYPE_ASCII only accepts ASCII text") from exc


def i16_pair_payload(dx: int, dy: int) -> bytes:
    return int(dx).to_bytes(2, "big", signed=True) + int(dy).to_bytes(2, "big", signed=True)


def u32_payload(value: int) -> bytes:
    return int(value).to_bytes(4, "big", signed=False)


def byte_payload(value: int) -> bytes:
    return bytes([int(value) & 0xFF])
